### PR TITLE
CA-396646: Fix bugs found from MRG

### PR DESCRIPTION
**Pull request by xueqingz**
test result :
```
ubuntu@ubuntu-server:~/code/terraform-provider-xenserver$ make testacc
if [ -z "/home/ubuntu/go/bin" ]; then echo "GOBIN is not set" && exit 1; fi
go mod tidy
go install .
ls -l /home/ubuntu/go/bin/terraform-provider-xenserver
-rwxrwxr-x 1 ubuntu ubuntu 23279409 Aug  6 03:16 /home/ubuntu/go/bin/terraform-provider-xenserver
source .env && TF_ACC=1 go test ./xenserver/ -v   -timeout 120m
=== RUN   TestAccNetworkDataSource
--- PASS: TestAccNetworkDataSource (0.67s)
=== RUN   TestAccVlanResource
--- PASS: TestAccVlanResource (2.97s)
=== RUN   TestAccNICDataSource
--- PASS: TestAccNICDataSource (0.62s)
=== RUN   TestAccPifDataSource
--- PASS: TestAccPifDataSource (1.08s)
=== RUN   TestAccSnapshotResource
--- PASS: TestAccSnapshotResource (5.66s)
=== RUN   TestAccSRDataSource
--- PASS: TestAccSRDataSource (0.63s)
=== RUN   TestAccNFSResource
--- PASS: TestAccNFSResource (4.32s)
=== RUN   TestAccSRResourceLocal
--- PASS: TestAccSRResourceLocal (4.22s)
=== RUN   TestAccSRResourceShared
--- PASS: TestAccSRResourceShared (3.67s)
=== RUN   TestAccVDIResource
--- PASS: TestAccVDIResource (5.04s)
=== RUN   TestAccVMDataSource
--- PASS: TestAccVMDataSource (2.31s)
=== RUN   TestAccVMResource
--- PASS: TestAccVMResource (4.16s)
=== RUN   TestAccLinuxVMResource
--- PASS: TestAccLinuxVMResource (2.10s)
PASS
ok      terraform-provider-xenserver/xenserver  37.444s
```

### CODE REVIEW

File: xenserver/vm_utils.go [hunk 1]
```diff
@@ -468,11 +470,11 @@ func getFirstTemplate(session *xenapi.Session, templateName string) (xenapi.VMRe
 
 	// Get the first VM template ref
 	for vmRef, record := range records {
-		if record.IsATemplate && strings.Contains(record.NameLabel, templateName) {
+		if record.IsATemplate && strings.Compare(record.NameLabel, templateName) == 0 {
```
Comment (acefei) [inline xenserver/vm_utils.go:473]: why not using equality operators (==) that is more straightforward and efficient approach for case-sensitive string comparison.

File: xenserver/vm_utils.go [hunk 2]
```diff
@@ -769,27 +777,42 @@ func updateVMMemory(session *xenapi.Session, vmRef xenapi.VMRef, data vmResource
 	return nil
 }
 
-func updateVMCPUs(session *xenapi.Session, vmRef xenapi.VMRef, plan vmResourceModel) error {
-	vcpus := int(plan.VCPUs.ValueInt32())
-	vcpusAtStartup, err := xenapi.VM.GetVCPUsAtStartup(session, vmRef)
+func updateVMMemory(ctx context.Context, session *xenapi.Session, vmRef xenapi.VMRef, plan vmResourceModel, state vmResourceModel) error {
+	staticMemMin, staticMemMax, dynamicMemMin, dynamicMemMax := getVMMemory(plan)
```
Comment (acefei) [inline xenserver/vm_utils.go:781]: using planStaticMemMin ... and stateStaticMemMin ... is more meaningful.
or just like what you did above to return a struct, then you could simply compare them by f() == g().
see: https://go.dev/play/p/zFXPqPjTtZ

File: xenserver/vbd_utils.go [hunk 3]
```diff
@@ -292,70 +312,67 @@ func getVDIUUIDFromISOName(session *xenapi.Session, isoName string) (string, err
 
 func setCDROM(ctx context.Context, session *xenapi.Session, vmRef xenapi.VMRef, plan vmResourceModel) error {
 	if plan.CDROM.IsUnknown() {
-		tflog.Debug(ctx, "---> CD-ROM is not set, use the default value from the VM template")
+		tflog.Debug(ctx, "---> CD-ROM is not set, use the default value")
```
Comment (acefei) [inline xenserver/vbd_utils.go:315]: the original message is more precise.

File: xenserver/vbd_utils.go [hunk 4]
```diff
@@ -174,13 +174,24 @@ func updateVBDs(ctx context.Context, plan vmResourceModel, state vmResourceModel
...
 
+	vmState, err := xenapi.VM.GetPowerState(session, vmRef)
+	if err != nil {
+		return errors.New(err.Error())
+	}
+
 	// Destroy VBDs that are not in plan
 	for vdiUUID, stateVBD := range stateHardDrivesMap {
 		if _, ok := planHardDrivesMap[vdiUUID]; !ok {
+			if vmState == xenapi.VMPowerStateRunning && stateVBD.Mode.ValueString() == "RW" {
```
Comment (acefei) [inline xenserver/vbd_utils.go:185]: When vm is running: 
1. RW VBD can't be removed, but RO VBD can.
2. RO VBD can't be created, but RW VBD can.
3. Both RW and RO VBD can't be updated.

Am I correct? CC @JinyuZhou-patty @xihuan-citrix

File: xenserver/vbd_utils.go [hunk 5]
```diff
@@ -174,13 +174,24 @@ func updateVBDs(ctx context.Context, plan vmResourceModel, state vmResourceModel
...
+				return errors.New("unable to delete the item with 'RW' mode in hard_drive for a running VM")
+			}
 			tflog.Debug(ctx, "---> Destroy VBD:	"+stateVBD.VBD.String())
 			err = xenapi.VBD.Destroy(session, xenapi.VBDRef(stateVBD.VBD.ValueString()))
 			if err != nil {
-				return errors.New(err.Error())
+				if !strings.Contains(err.Error(), "HANDLE_INVALID") {
```
[the remark this answers is not in the document]
Comment (xueqingz) [reply xenserver/vbd_utils.go:191]: Handle invalid means the vbdRef is invalid at that time.

File: xenserver/vbd_utils.go (hunk 3 above)
Comment (xueqingz) [reply xenserver/vbd_utils.go:315]: It's changed, now used for both setting and updating , not always from template

File: xenserver/vbd_utils.go (hunk 4 above)
Comment (xueqingz) [reply xenserver/vbd_utils.go:185]: 1. right 
2. right
3. Both RW an RO, the attribute 'mode' and 'bootable' can't be updated.

File: xenserver/vbd_utils.go (hunk 3 above)
Comment (acefei) [reply xenserver/vbd_utils.go:315]: if updating, CDROM.IsUnknown() means the cdrom attribute was removed, in this case, where does the default value comes from?

File: xenserver/vbd_utils.go (hunk 4 above)
Comment (acefei) [reply xenserver/vbd_utils.go:185]: Hope QA have the related cases to cover this changes.

File: xenserver/vbd_utils.go (hunk 3 above)
Comment (xueqingz) [reply xenserver/vbd_utils.go:315]: tf state is no changed.

**Review by acefei — APPROVED** [review]

**Review by xihuan-citrix — APPROVED** [review]

**Pull request merged**